### DISCUSSION TITLE
Is there the generate_db_auth_token in SOTO?

### DISCUSSION THREAD

**Discussion by alteredtech** [soto-project/soto#670] [Q&A] [answered]
I am trying pet project to convert our tools to swift. One thing I have not been able to find is the ability to generate a db auth token that can be used to sign into RDS with a mysql client. 
I have tried to mimic it manually with swift but I realized that the access key was not the same as my access key.

**Comment by adam-fowler** [comment]
No there isn't a version of this function, but it should be fairly easy to implement as it uses SIGV4 to create it.
The java v2 sdk code is here https://github.com/aws/aws-sdk-java-v2/blob/b0f57372dbe6be1fb2a6a54b05ad1f1a7e1e7700/services/rds/src/main/java/software/amazon/awssdk/services/rds/DefaultRdsUtilities.java#L67

From what I can work out it creates a URL `https://<RDS hostname>:<port>/?Action=connect&DBUser=<username>` which it then passes through the signer using `rds-db` as the signing name
```swift
import SotoSignerV4

let url = "https://<_RDSHOSTNAME_>:<_RDSPORT_>/?Action=connect&DBUser=<username>"
let credentials: Credential = StaticCredential(
    accessKeyId: "_MYACCESSKEY_", 
    secretAccessKey: "_MYSECRETACCESSKEY_"
)
let signer = AWSSigner(credentials: credentials, name: "rds-db", region: "_MYREGION_")
let signedURL = signer.signURL(
    url: url,
    method: .GET
)
// drop "https://"
let authToken = signedURL.absoluteString.dropFirst(8)
```
You need to obviously replace the hostname, port, credentials and region in the code above. If you already have a AWSClient setup you can use `let credentials = try await awsClient.credentialProvider.getCredential()` to get your credentials.

This is completely untested code. I would most appreciate it if you replied whether it works for you or not, then I can stick it in the documentation.

**Comment by alteredtech** [comment, marked as answer]
Ah you beautiful person. Thank you tons! Here is my working code. You got pretty much all of it.

```
    let region = "your_region"
    let serverHost = "RDS_host"
    let serverPort = "RDS_port"
    let dbUser = "DB_user"
    let secretKey = "your_secret)key"
    let accessKey = "your_access_key"
    let sessionToken = "your_session_token"
    let url = "https://\(serverHost):\(serverPort)/?Action=connect&DBUser=\(dbUser)"
    let credentials: Credential = StaticCredential(
        accessKeyId: accessKey,
        secretAccessKey: secretKey,
        sessionToken: sessionToken
    )
    let signer = AWSSigner(credentials: credentials, name: "rds-db", region: region)
    let signedURL = signer.signURL(
        url: URL(string: url)!, //I forced unwrapped here
        method: .GET,
        expires: TimeAmount(Duration.seconds(900))
    )
    let authToken = signedURL.absoluteString.dropFirst(8)
```

**Reply by adam-fowler** [reply]
Thanks, I'll get that added to the documentation on soto.codes
